File: algorithms/121.c
#include <stdio.h>

int maxProfit(int* prices, int pricesSize)
{
    if ( pricesSize < 2 )
        return 0;
    int i;
    int buy = pricesSize-2, sell = pricesSize-1;
    int max = sell;
    for ( i = pricesSize-2; i >= 0; i-- )
    {
        if ( prices[max]-prices[i] > prices[sell]-prices[buy] )
            buy = i, sell = max;
        if ( prices[i] > prices[max] )
            max = i;
    }
    return prices[sell]>prices[buy] ? prices[sell]-prices[buy] : 0;
}

int main()
{
    int arr[] = {5,1,4,7,24};
    printf("%d\n", maxProfit(arr, sizeof(arr)/sizeof(arr[0])));
    return 0;
}

File: algorithms/python/121.py
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        if not prices: return 0
        minP = prices[0]
        rls = 0
        for i, n in enumerate(prices):
            rls = max(rls, n-minP)
            minP = min(minP, n)
        return rls
